physx/src/physics.rs

```diff
@@ -9,34 +9,115 @@ Wrapper interface for Physics
+    owner::Owner,
+    pruning_structure::PruningStructure,
+    rigid_actor::RigidActor,
+    rigid_dynamic::RigidDynamic,
+    rigid_static::RigidStatic,
+    scene::{Scene, SceneDescriptor},
+    shape::{Shape, ShapeFlags},
+    traits::{Class, PxFlags},
+    triangle_mesh::TriangleMesh,
+    visual_debugger::VisualDebugger,
 };
 
-use glam::{Mat4, Vec3};
-use physx_sys::*;
-use std::ptr::null_mut;
+use std::{marker::PhantomData, ptr::null_mut};
+
+use physx_sys::{
+    phys_PxCloseExtensions,
+    phys_PxCreatePhysics,
+    // TODO implement the extensions interface, move these there isntead of here?
+    //phys_PxCreateBasePhysics,  used with extentions
+    phys_PxInitExtensions,
+    physx_create_physics, // TODO remove this? it just calls through to PxCreatePhysics
+    FilterShaderCallbackInfo,
+    PxConstraintConnector,
+    PxConstraintShaderTable,
+    PxFilterFlag,
+    // TODO write wrappers and import from them instead
+    PxInputStream,
+    PxPairFlag,
+    PxPhysicsInsertionCallback,
+    // TODO unify creation into `fn create<D: Descriptor<T>>(descriptor: &D) -> Option<&mut T>`
+    // or `fn create<T>(descriptor: Descriptor<T>) -> Option<&mut T>`
+    PxPhysics_createAggregate_mut,
+    PxPhysics_createArticulationReducedCoordinate_mut,
+    PxPhysics_createArticulation_mut,
+    PxPhysics_createBVHStructure_mut,
+    PxPhysics_createConstraint_mut,
+    PxPhysics_createConvexMesh_mut,
+    PxPhysics_createHeightField_mut,
+    PxPhysics_createMaterial_mut,
+    PxPhysics_createPruningStructure_mut,
+    PxPhysics_createRigidDynamic_mut,
+    PxPhysics_createRigidStatic_mut,
+    PxPhysics_createScene_mut,
+    PxPhysics_createShape_mut_1,
+    PxPhysics_createTriangleMesh_mut,
+    PxPhysics_getBVHStructures,
+    PxPhysics_getConvexMeshes,
+    PxPhysics_getHeightFields,
+    PxPhysics_getMaterials,
+    PxPhysics_getNbBVHStructures,
+    PxPhysics_getNbConvexMeshes,
+    PxPhysics_getNbHeightFields,
+    PxPhysics_getNbMaterials,
+    PxPhysics_getNbScenes,
+    PxPhysics_getNbShapes,
+    PxPhysics_getNbTriangleMeshes,
+    PxPhysics_getPhysicsInsertionCallback_mut,
+    PxPhysics_getScenes,
+    PxPhysics_getShapes,
+    PxPhysics_getTolerancesScale,
+    PxPhysics_getTriangleMeshes,
+    PxPhysics_release_mut,
+    //PxPhysics_registerDeletionListener_mut,
+    //PxPhysics_registerDeletionListenerObjects_mut,
+    //PxPhysics_unregisterDeletionListener_mut,
+    //PxPhysics_unregisterDeletionListenerObjects_mut,
+    //PxPhysics_getFoundation_mut, // probably not necessary? also sketch wrt obrm
+    // PxPhysics_createShape_mut, PxPhysics_createShape_mut_1 is the same but more
+    // TODO move PxTolerancesScale, callbacks and ContactPair stuff to their own files?
+    PxTolerancesScale,
+    PxTolerancesScale_new,
+};
 
-pub const PX_PHYSICS_VERSION: u32 = (4 << 24) + (1 << 16);
+pub const PX_PHYSICS_VERSION: u32 = crate::version(4, 1, 1);
 
+/*
 #[no_mangle]
 pub extern "C" fn on_contact_callback(
```

Comment:
The simulation event callbacks are now traits.

I removed the `simulation_filter_shader` function as well, the API there is now passing in a function pointer of type `physx_sys::SimulationFilterShader` wrapped in an enum for only-default/default-first/no-default during scene creation.  That function is that type, so if it is useful as a Rust default it could be re-added easily.